### ISSUE TITLE
ClassCastException using ByteBuffer as key for ChronicleMap

### ISSUE THREAD

**Issue by jackmalkovick** [OpenHFT/Chronicle-Map#181]
I tried a very simple example where I try to write int -> ByteBuffer and reuse the array while reading the values from the map.

```
try (ChronicleMap<Integer, ByteBuffer> map = ChronicleMap
                    .of(Integer.class, ByteBuffer.class)
                    .entries(1)
                    .averageValueSize(30)
                    .createPersistedTo(new File("e:/temp/test.obj"))) {

                byte[] out = new byte[] { 1, 2, 3};         
                map.put(1, ByteBuffer.wrap(out)); // ClassCastException

                ByteBuffer buff = ByteBuffer.allocate(30);
                ByteBuffer in = map.getUsing(1, buff); // hoping that 'ByteBuffer in'
                                                       // is redundant
    }
```

Unfortunately I get an unexpected `ClassCastException`, because Chronicle expected a `byte[]`

```
Exception in thread "main" java.lang.ClassCastException: ChronicleMap{name=null, file=E:\temp\test.obj, identityHashCode=809762318}: Value must be a [B but was a class java.nio.HeapByteBuffer
    at net.openhft.chronicle.map.VanillaChronicleMap.checkValue(VanillaChronicleMap.java:190)
    at net.openhft.chronicle.map.VanillaChronicleMap.put(VanillaChronicleMap.java:707)
    at test.SerTest.main(SerTest.java:28)
```

**Comment by RobAustin** [comment]
```
package software.chronicle.enterprise.map.queue.backed.map;

import net.openhft.chronicle.bytes.Bytes;
import net.openhft.chronicle.hash.serialization.SizedReader;
import net.openhft.chronicle.hash.serialization.SizedWriter;
import net.openhft.chronicle.map.ChronicleMap;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.File;
import java.io.IOException;
import java.nio.ByteBuffer;

/**
 * Created by Rob Austin
 */
public class Example {

    static class ByteBufferMarshaller implements SizedReader<ByteBuffer>, SizedWriter<ByteBuffer> {

        @NotNull
        @Override
        public ByteBuffer read(final Bytes in, final long size, @Nullable final ByteBuffer using) {
            ByteBuffer u;
            if (using != null) {
                using.position(0).limit((int)size);
                u = using;
            } else {
                u = ByteBuffer.allocate((int) size);
            }
            while(in.readRemaining()>0 && u.position()< u.limit()) {
                u.put(in.readByte());
            }
            return u;
        }

        @Override
        public long size(@NotNull final ByteBuffer toWrite) {
            return toWrite.limit() - toWrite.position();
        }

        @Override
        public void write(final Bytes out, final long size, @NotNull final ByteBuffer toWrite) {
            while(toWrite.position() < toWrite.limit()) {
                out.writeByte(toWrite.get());
            }

        }
    }

    public static void main(String[] args) throws IOException {
        try (ChronicleMap<Integer, ByteBuffer> map = ChronicleMap
                .of(Integer.class, ByteBuffer.class)
                .entries(1)
                .averageValueSize(30)
                .valueMarshaller(new ByteBufferMarshaller())
                .createPersistedTo(new File("test.obj"))) {

            byte[] out = new byte[]{1, 2, 3};
            ByteBuffer wrap = ByteBuffer.wrap(out);
            map.put(1, wrap); // ClassCastException

            ByteBuffer buff = ByteBuffer.allocate(30);
            buff.clear();
            ByteBuffer in = map.getUsing(1, buff); // hoping that 'ByteBuffer in'

            System.out.println(in.toString());
            // is redundant
        }
    }
}
```

**Comment by hft-team-city** [comment]
Released in [Chronicle-Map-3.17.3](https://github.com/OpenHFT/Chronicle-Map/releases/chronicle-map-3.17.3), [BOM-2.17.262](https://github.com/OpenHFT/OpenHFT/releases/chronicle-bom-2.17.262)
